xlators/cluster/dht/src/dht-inode-write.c

```diff
@@ -10,71 +10,141 @@
-static int
-dht_discard2(xlator_t *this, xlator_t *subvol, call_frame_t *frame, int ret);
-static int
-dht_zerofill2(xlator_t *this, xlator_t *subvol, call_frame_t *frame, int ret);
+void
+dht_inode_write_wind(call_frame_t *frame, xlator_t *wind_subvol);
 
 int
-dht_writev_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int op_ret,
-               int op_errno, struct iatt *prebuf, struct iatt *postbuf,
-               dict_t *xdata)
+dht_inode_write2(xlator_t *this, xlator_t *subvol, call_frame_t *frame,
+                 int ret);
+
+void
+dht_inode_write_unwind(glusterfs_fop_t fop, call_frame_t *frame, int32_t op_ret,
+                       int32_t op_errno, struct iatt *prestat,
+                       struct iatt *poststat, dict_t *xdata)
+{
+    switch (fop) {
+        case GF_FOP_WRITE:
+            DHT_STACK_UNWIND(writev, frame, op_ret, op_errno, prestat, poststat,
+                             xdata);
+            break;
+
+        case GF_FOP_ZEROFILL:
+            DHT_STACK_UNWIND(zerofill, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_DISCARD:
+            DHT_STACK_UNWIND(discard, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_TRUNCATE:
+            DHT_STACK_UNWIND(truncate, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_FTRUNCATE:
+            DHT_STACK_UNWIND(ftruncate, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_FALLOCATE:
+            DHT_STACK_UNWIND(fallocate, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+            /*TODO: too many cbk for dirs is different, cbk for mds is
+             * different*/
+        case GF_FOP_FSETATTR:
+            DHT_STACK_UNWIND(fsetattr, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_SETATTR:
+            DHT_STACK_UNWIND(setattr, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+        default:
+            GF_ASSERT(0);
+            break;
+    }
+}
+
+int
+dht_set_protect_from_external_writes(call_frame_t *frame)
+{
+    dht_local_t *local = frame->local;
+    xlator_t *this = frame->this;
+    int ret = 0;
+    if (local->fop == GF_FOP_WRITE) {
+        if (!local->xattr_req) {
+            local->xattr_req = dict_new();
+            if (!local->xattr_req) {
+                ret = -ENOMEM;
+                goto out;
+            }
+        }
+
+        ret = dict_set_uint32(local->xattr_req, GF_PROTECT_FROM_EXTERNAL_WRITES,
+                              1);
+        if (ret) {
+            gf_msg(this->name, GF_LOG_ERROR, DHT_MSG_DICT_SET_FAILED, 0,
```

Comment:
nit: line 76 can be removed if you use 'THIS->name' instead?

Reply:
And then you can probably pass 'local' to the function instead of frame.

Reply:
Passing frame to include both local, this is a standard practice. So I am not keen on changing this.